bug: Fix Aliased dependencies condition not respecting specified default value

oliviermichaelis opened the pull request:
This PR fixes #9052.

**What this PR does / why we need it**:
In case a chart is referenced multiple times as a dependency with a condition, a `false` condition is only respected for the dependency that is processed first. The remaining dependencies whose condition is `false` will get rendered incorrectly.

I found out that the chart dependency`Metadata` is initialized once the dependency`Chart` is loaded (e.g. from the filesystem). This `Metadata` object is then referenced by all aliased depency `Chart`.

Once the first dependency is processed, the `Metadata`s `Name` is changed to the `Alias`:
https://github.com/helm/helm/blob/960b3b53c6c30d72c68738fbed3ccf3724b30130/pkg/chartutil/dependencies.go#L143-L145

For the next dependency, the check in Line 132 will fail since `req.Name` has now changed to the alias:
https://github.com/helm/helm/blob/960b3b53c6c30d72c68738fbed3ccf3724b30130/pkg/chartutil/dependencies.go#L130-L135

The remaining disabled dependencies will therefore not get removed from the list of charts that will be rendered.

I decided to deep-copy the`Metadata`s dependencies in this PR to prevent metadata corruption.

**Special notes for your reviewer**:
This PR contains a test case for a chart that renders incorrectly with the current `master`, but renders correctly with this PR.
The structure of the chart is:
```
chart-with-dependency-aliased-twice-with-condition:
  childFoo:
    grandchildAlias
  childBar:
    grandchildAlias
```

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

oliviermichaelis committed:
fix(pkg/chartutil) Remove disabled dependencies

Remove aliased dependencies that are disabled with a condition.
Before, if a chart had multiple aliased dependencies that reference the
same chart, the metadata was shared between all dependencies. This led
to dependencies not being processed correctly and resulted in a bug
where disabled dependencies were not removed and therefore rendered.

The metadatas dependencies are now deep-copied to prevent metadata
corruption.

Signed-off-by: Olivier Michaelis <38879457+oliviermichaelis@users.noreply.github.com>

oliviermichaelis commented:
The CircleCI build did not run, is it possible to retrigger that?

bacongobbler commented:
We've seen this occur before when other community members fork Helm and point their fork at another CircleCI account IIRC. I ca't fully recall what's the correct course of action here, but looking at the webhook system from GIthub's end it looks like the webhook payloads are being sent to CircleCI just fine on our end.

Can  you please look into this on your end?

oliviermichaelis commented:
Thanks for the pointer, I have deleted the CircleCI integration on my fork. Can we somehow trigger the build again?

msvechla commented:
@bacongobbler we also experience this bug, can we get this merged?

FireDrunk commented:
I've just stumbled across this bug, and this seems fixed in the latest version if this is the issue you were getting at:

```yaml
dependencies:
  # Databases
  - name: postgresql
    version: 12.8.0
    repository: oci://registry-1.docker.io/bitnamicharts
    alias: db-a
    condition: db-a.enabled

  - name: postgresql
    version: 12.8.0
    repository: oci://registry-1.docker.io/bitnamicharts
    alias: db-b
    condition: db-b.enabled
```

This works.

joejulian commented:
Tested your repro with helm 3.12.2. Still a problem.

pderrierw commented:
Please can we have this merged ? This is a real issue for us

we have a helm chart A with a conditional dependency on redis:
`dependencies:
  - name: redis
    version: 17.3.8
    condition: redis.enabled
`

in the application we reference multiple time the chart A as dependency:
`dependencies:
  - name: A
    alias: application1
  - name: A
    alias: application2
`

with following values.yaml:
`application1:
  redis:
    enabled: false
    
application2:
  redis:
     enabled: true`
     
     
Impossible to get a stable rendering of the final chart, sometimes the redis is rendered, sometimes not, I have the impression it depends on the name of application1 and 2 (so of the ordering of the processing by helm)

That matches the content of this MR, thanks @oliviermichaelis for the fix
Why isn't merged yet?

scottrigby commented:
The code does work, but not sure that `TestChartWithDependencyAliasedTwiceWithCondition()` is testing this fully. Here's what a manual test gives:

1. first, we need to update the test chart included in this PR to:
    ```diff
    $ git diff pkg/chartutil/testdata/chart-with-dependency-aliased-twice-with-condition/charts/child/charts/grandchild/templates/service.yaml
    ...
    @@ -1,7 +1,7 @@
     apiVersion: v1
     kind: Service
     metadata:
    -  name: {{ include "grandchild.fullname" . }}
    +  name: {{ include "grandchild.fullname" . | lower }}
       labels:
         {{- include "grandchild.labels" . | nindent 4 }}
     spec:
    ```

    otherwise it will fail with:
    ```console
    $ helm install test pkg/chartutil/testdata/chart-with-dependency-aliased-twice-with-condition \
      --set childFoo.grandchildAlias.enabled=true
    Error: INSTALLATION FAILED: 1 error occurred:
	    * Service "test-grandchildAlias" is invalid: metadata.name: Invalid value: "test-grandchildAlias": a DNS-1035 label must consist of lower case alphanumeric characters or '-', start with an alphabetic character, and end with an alphanumeric character (e.g. 'my-name',  or 'abc-123', regex used for validation is '[a-z]([-a-z0-9]*[a-z0-9])?')
    ```
    because while Helm has `ValidateReleaseName()` in pkg/chartutil/validate_name.go`, the dependency alias when passed to `.Chart.Name` doesn't validate on install/upgrade, so (for now) we can only test by converting to lower in the template before attempting to create a service with an uppercase character ("grandchildAlias").

2. now we can reproduce the error:

    ```console
    $ helm install test pkg/chartutil/testdata/chart-with-dependency-aliased-twice-with-condition \
      --set childFoo.grandchildAlias.enabled=true
      NAME: test
      STATUS: deployed
      ...
    
    $ k get all
    NAME                            TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)   AGE
    service/test-grandchild        ClusterIP   10.96.7.47      <none>        80/TCP    12s
    service/test-grandchildalias   ClusterIP   10.96.162.130   <none>        80/TCP    12s
    ```
    I see the error is there is one dependency chart's template installed—using the alias. And another installed using the dependency name—not the alias. Only the first one should have been created.
    
    If you change the order, and only enable the 2nd dependency in the list `--set childBar.grandchildAlias.enabled=true` instead, you get only the dependency chart's template installed using the dependency name—not the alias:
    ```console
    $ helm uninstall test
    $ helm install test pkg/chartutil/testdata/chart-with-dependency-aliased-twice-with-condition \
      --set childBar.grandchildAlias.enabled=true
    NAME                      TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
    service/test-grandchild   ClusterIP   10.96.190.22   <none>        80/TCP    3s
    ```
    
3. Test this PR. Seems to fix it:
    ```console
    $ helm uninstall test
    $ gh pr checkout 9184
    $ make build
    # test first case above
    $ ./bin/helm install test pkg/chartutil/testdata/chart-with-dependency-aliased-twice-with-condition \
      --set childFoo.grandchildAlias.enabled=true
      ...
      
    $ k get all
    NAME                           TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
    service/test-grandchildalias   ClusterIP   10.96.73.213   <none>        80/TCP    2s
    
    # test second case above
    $ ./bin/helm install test pkg/chartutil/testdata/chart-with-dependency-aliased-twice-with-condition \
      --set childBar.grandchildAlias.enabled=true
    $ k get all
    NAME                           TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
    service/test-grandchildalias   ClusterIP   10.96.57.100   <none>        80/TCP    3s
    ```
    
4. Fun note, if you want to test enabling both dependencies, you'll need to do a name override on one of the grandchildren charts, otherwise the service will fail creation because the same name will be used and `service/test-grandchildalias` already exists. Like this:
    ```console
    $ helm install pkg/chartutil/testdata/chart-with-dependency-aliased-twice-with-condition \
      --set childFoo.grandchildAlias.enabled=true \
      --set childBar.grandchildAlias.enabled=true \
      --set childBar.grandchildAlias.nameOverride=bar-grandchild
    $ k get all
                  TYPE        CLUSTER-IP    EXTERNAL-IP   PORT(S)   AGE
    service/test-bar-grandchild    ClusterIP   10.96.23.11   <none>        80/TCP    2s
    service/test-grandchildalias   ClusterIP   10.96.86.83   <none>        80/TCP    2s
    ```

So good work on the code change. Correct me if I'm wrong about the automated test—not sure it covers this fully.

oliviermichaelis committed:
fix: convert to resource name to lower case to comply with kubernetes validation

Signed-off-by: Olivier Michaelis <38879457+oliviermichaelis@users.noreply.github.com>

oliviermichaelis commented:
Thanks for your effort @scottrigby, much appreciated! I incorporated your feedback about the invalid resource name in a1b5d6f0222247589dfdfa6c31257a9ab9b856e2.

>Correct me if I'm wrong about the automated test—not sure it covers this fully.

Can you be more specific about this statement, please? What exactly is missing? This is the first time I'm looking at my code from a couple of years ago and I'm certainly missing something. Thanks!

scottrigby commented:
> Thanks for your effort @scottrigby, much appreciated! I incorporated your feedback about the invalid resource name in [ffd5d61](https://github.com/helm/helm/commit/ffd5d6106f88aa93dbe0705be6ef6f18550532f8).
> 
> > Correct me if I'm wrong about the automated test—not sure it covers this fully.
> 
> Can you be more specific about this statement, please? What exactly is missing? This is the first time I'm looking at my code from a couple of years ago and I'm certainly missing something. Thanks!

What I meant was that if the unit test didn't catch the template bug I found, then why are we adding templating in the test chart.
You're right that for this single use case/bugfix, the test covers it ok. We are processing only the dependencies.
Given that, can you please simplify the testdata chart to have only the bare minimum needed for the test itself? It doesn't need boilerplate comments, values, or even templating. That will make it more clear what the test chart is for when paired with the unit test. And then I think this will be good to go.

oliviermichaelis committed:
fix(pkg/chartutil): cleaning up to address review comments

- shorten testdata chart name
- remove unnecessary and distracting files, comments, and values

Signed-off-by: Olivier Michaelis <38879457+oliviermichaelis@users.noreply.github.com>

oliviermichaelis commented:
@scottrigby I rebased on the latest `main` and addressed your comments by:
- Renamed/shortened testdata chart name
- Removed boilerplate comments, values, and templating to the absolute minimum. This should make it easier to understand what's actually going on in the test

scottrigby reviewed (approved):
@oliviermichaelis hi again—I was able to review your changes to this PR since you made them in February, and they look good. Are you able to fix the conflicts since changes were made in main? LGTM once that's complete

hobti01 commented:
Can confirm that this change fixes the issue when applied to the 3.17 codebase.  We're very much looking forward to seeing this long-standing issue squashed.

oliviermichaelis commented:
In the meantime #9175 for v4 got merged and there is also #30802 to backport the fix to v3. As far as I can tell, they both address the same problem as this PR.
Since a definitive solution seems close,I would consider to close my PR in favor of the previously mentioned PRs.
